### PR TITLE
RGB_Strip, RGBW_Strip, and LEDSTRIP (WS2812B) Support

### PR DESCRIPTION
**Pull request by ghost**
RGB Strip support by @boweeble.
Modified from PR #5395 to include working outside of SD printing.
Added option set the timeout duration.
Added RGBW strip support.
Added addressable RGB strip support.
Moved all LED related stuff to `configuration_adv.h`.
LCD option for the addressable strip doesn't fully work yet. And am trying to get it to work for all LED types.

### CODE REVIEW

**Comment by cdedwards** [comment]
```diff
+
+      // This variant uses 3 separate pins for the RGB components.
+      // If the pins can do PWM then their intensity will be set.
+      digitalWrite(RGB_LED_R_PIN, r ? HIGH : LOW);
+      digitalWrite(RGB_LED_G_PIN, g ? HIGH : LOW);
+      digitalWrite(RGB_LED_B_PIN, b ? HIGH : LOW);
+      analogWrite(RGB_LED_R_PIN, r);
+      analogWrite(RGB_LED_G_PIN, g);
+      analogWrite(RGB_LED_B_PIN, b);
+
+    #endif
```
think you are missing a #ifdef #else #endif in here. rgb_strip.cpp

**Comment by ghost** [comment]
I don't believe that is an issue being that it's in the RGB_Strip.cpp file and not another file.

**Comment by Blue-Marlin** [comment]
Where do you want to go?
Looks a bit old fashioned - like https://www.youtube.com/watch?v=LUID0jSh2Ic
For a real disco Marlin needs to be a DMX master these days.

**Comment by ghost** [comment]
Lol

**Comment by 8mccm8** [comment]
Hi @boweeble,

I tried to see the PWM signal on D4 D5 D6 with my scop without success!
D4/D5/D6 are open drain pins ?
PWM freq is 500kHz?

File: Marlin/example_configurations/Cartesio/Configuration.h [hunk 1]
```diff
@@ -1459,5 +1459,43 @@
...
+  //=========================== Printer Event LEDs ============================
+  //===========================================================================
+ *  When printing, the LEDs will display printer status
+ *  - LEDs will gradually change from blue to violet as the heated bed gets to target temp
+ *  - LEDs will gradually change from violet to red as the hotend gets to temperature
+ *  - LEDs will change to white to illuminate work surface
+ *  - LEDs will change to green once print has finished
+ *  - LEDs will turn off after the print has finished and the user has pushed a button 
+ * */
+#if ENABLED(BLINKM) || ENABLED(RGB_LED) || ENABLED(RGB_STRIP)
```
Comment (8mccm8) [inline Marlin/example_configurations/Cartesio/Configuration.h:1564]: hi where is the RGB_STRIP definition variable? thx.

Comment (ghost) [reply Marlin/example_configurations/Cartesio/Configuration.h:1564]: Line 1478.

Comment (8mccm8) [reply Marlin/example_configurations/Cartesio/Configuration.h:1564]: why I can't see this variable in the last RCBugFix?

Comment (ghost) [reply Marlin/example_configurations/Cartesio/Configuration.h:1564]: This PR has not been merged yet. So, it is not part of RCBugFix....yet.

Comment (8mccm8) [reply Marlin/example_configurations/Cartesio/Configuration.h:1564]: ok thanks understood newbie powered :)

Comment (ghost) [reply Marlin/example_configurations/Cartesio/Configuration.h:1564]: It's all good. :)

**Comment by 8mccm8** [comment]
hi @boweeble and @Tannoo,
I tested this new feature, for me is not working correctly:

turn on print => stripe is OFF
start printing from SD: 
heating => strip is full RED
heating ok => strip goes green

haven't seen any fading... will go in this code later if you have some clue....

**Comment by ghost** [comment]
@8mccm8,
Do all colors work correctly with M150?
Main colors to test are RED, GREEN, and BLUE.
Test intensity also.

**Comment by ghost** [comment]
Intensity, meaning: M150 R127 or M150 B10 or M150 R10 U10 B10.

**Comment by ghost** [comment]
I added more events that will have to wait for the new M600 PR (#5737) to be merged.

**Comment by 8mccm8** [comment]
@Tannoo yep M150 working well :)

**Comment by teemuatlut** [comment]
Any chance to extend this to RGBW?

**Comment by ghost** [comment]
>Any chance to extend this to RGBW?

The non-addressable type, yes.

**Comment by ghost** [comment]
@teemuatlut, test it now. I do not have an RGBW strip to test it with. I will order one though.

**Comment by teemuatlut** [comment]
Thank you for adding the option. I just ordered mine last night.

**Comment by ghost** [comment]
Let me know if it works. I have an RGBW strip on order also for testing. I might even change my strip out. lol

**Comment by ghost** [comment]
I got the strip I ordered in and tested it. Works as intended.

**Comment by nesalkiN** [comment]
I have addressable led strips with ws2812b.
Could that be used with Marlin?
http://3dprinters.proboards.com/thread/21/neopixel-ws2812-ledstrip-light-printer
What "#define LEDSTRIP" does the author talk about?

Edit: Ah, in the fork:
https://github.com/phil-hx/Marlin
Made some cut and pastes and got my led strip working. I can't say that I know what I am doing. ;)
Using 1 pin (Y_MAX) and GND connected in line with the led strip (using separate power supply for the strip). Changed V to U to match BLINKM and RGB_LED for M150.

**Comment by ghost** [comment]
I am going to check that out.

**Comment by nesalkiN** [comment]
PRINTER_EVENT_LEDS with support for ws2812b would be great! I have ~1 meter strip with 25 leds not used (got 5 meter for my diy "ambilight") , I can send it to you if you are interested.

Edit: I noticed your Tail-Light-Strip repo. Adafruit's NeoPixel's use either WS2812B or SK6812, same protocol and 5050 RGB leds so you may already have some? ;)

**Comment by ghost** [comment]
I have some left over from another project. They are just rgb though.

I got that same code inserted into mine and it compiles with the option off. I am going to test it soon.. Yes, I will get the printer events working with it also.

**Comment by nesalkiN** [comment]
Yeah, no w. Just need the proper mixture of red, green and blue.  ;)

**Comment by ghost** [comment]
>Edit: I noticed your Tail-Light-Strip repo. Adafruit's NeoPixel's use either WS2812B or SK6812, same protocol and 5050 RGB leds so you may already have some? ;)

Yes, that would be that project.

**Comment by ghost** [comment]
@NiklasOlofsson, what library did you use and where did you get it?

**Comment by nesalkiN** [comment]
The code in this fork https://github.com/phil-hx/Marlin use the FastLED library added from the manager.

**Comment by ghost** [comment]
I used that one also.
I can't get my strip to do anything except from interference. I have even used their RC3 version of Marlin and still will not work. I'm guessing that I'm using the wrong fastLED library.

Well, I got it to work with their RC3. I don't know why they were missing code from marlin_main.cpp. I got it.

**Comment by ghost** [comment]
Now, to make it work for RC8-RCBugFix.

**Comment by nesalkiN** [comment]
I modified the latest RCBugFix-code with code from that repo and my led strip works well.

**Comment by ghost** [comment]
https://github.com/FastLED/FastLED.git is where I got the library.

**Comment by nesalkiN** [comment]
Looks to be the same. I installed using the library manager, searched for fastled.
Copied ledstrip.cpp and ledstrip.h from repo. Added the code to Marlin_main.cpp and this to my configuration:
```
> #define LEDSTRIP
> 
> #if ENABLED(LEDSTRIP)
>   #define LEDSTRIP_PIN 28 // Y_MAX on GT2560
>   #define LEDSTRIP_NLED 25 
>   #define LEDSTRIP_NSEGMENT 3 // LEDS may be divided in segments for ease of use
>   #define LEDSTRIP_EXCHANGE_RU //If you need to send green, red, blue instead of red, green, blue
>   #define LEDSTRIP_TYPE WS2812B
> #endif
```
I needed the "LEDSTRIP_EXCHANGE_RU" (changed RV to RU to match the R-U-B naming for the already present code for BLINKM and RGB_LED, see ledstrip.cpp) enabled because it displayed green when it should show red.

**Comment by ghost** [comment]
Yes, I got it working on the RC3 because I was having some issues with RCBugFix. I'm porting over now.

**Comment by teemuatlut** [comment]
FYI: FastLed does not yet support RGBW lights, unlike the Adafruit's NeoPixel library. But it's in the works and I think Marlin should still rather use the faster library.

**Comment by ghost** [comment]
I got it working by M150, but the LCD options just return:
```
echo:enqueueing "me print"
echo:Unknown command: "me print"
```
I'm not sure why yet. The RC3 version worked fine.

**Comment by ghost** [comment]
>I needed the "LEDSTRIP_EXCHANGE_RU" (changed RV to RU to match the R-U-B naming for the already present code for BLINKM and RGB_LED, see ledstrip.cpp) enabled because it displayed green when it should show red.

I had to do the same.

**Comment by nesalkiN** [comment]
WS2812B need G-R-B, sent in that order. :)
me print? I think I have seen it before but I have only used the code to control the leds by using M150.

**Comment by ghost** [comment]
I changed `enqueue_and_echo_command_P(ledstripgcode);` to `enqueue_and_echo_command(ledstripgcode);` and it now works.

**Comment by ghost** [comment]
I got the M150 working. I am now working and testing the printer events with it.

**Comment by boweeble** [comment]
@Tannoo - Awesome job of picking up the ball and running with it!  I've been busy with life and have not had anytime to play around with this since I first implemented it.  Love what you've done so far!  Will be installing your latest on my printer tonight...

**Comment by ghost** [comment]
@boweeble,
Thank you. You are more than welcome to submit corrections and/or additions. As I am sure that my code is not the greatest. The LCD options don't turn anything on, but will turn it off. I'm still looking into that as a lower priority. My higher priorities were to get the ledstrip code working with M150 and the printer events.

I am still testing. Something weird is going on with my z axis raise after homing.

**Comment by ghost** [comment]
Travis may have passed, but this:
```
The command "opt_enable_adv AUTOMATIC_CURRENT_CONTROL STEALTHCHOP" exited with 0.
8.21s
$ build_marlin
Picked up JAVA_TOOL_OPTIONS: 
Loading configuration...
Xlib:  extension "RANDR" missing on display ":1.0".
Initializing packages...
Preparing boards...
Verifying...
sketch/Marlin_main.cpp: In function 'void automatic_current_control(const TMC2130Stepper&)':
sketch/Marlin_main.cpp:10407:39: warning: passing 'const TMC2130Stepper' as 'this' argument of 'bool TMC2130Stepper::checkOT()' discards qualifiers [-fpermissive]
       const bool is_otpw = st.checkOT(), // Check otpw even if we don't adjust. Allows for flag inspection.
                                       ^
sketch/Marlin_main.cpp:10408:49: warning: passing 'const TMC2130Stepper' as 'this' argument of 'bool TMC2130Stepper::getOTPW()' discards qualifiers [-fpermissive]
                  is_otpw_triggered = st.getOTPW();
                                                 ^
sketch/Marlin_main.cpp:10412:48: warning: passing 'const TMC2130Stepper' as 'this' argument of 'uint16_t TMC2130Stepper::getCurrent()' discards qualifiers [-fpermissive]
         const uint16_t current = st.getCurrent() + CURRENT_STEP;
                                                ^
sketch/Marlin_main.cpp:10413:71: warning: passing 'const TMC2130Stepper' as 'this' argument of 'void TMC2130Stepper::SilentStepStick2130(uint16_t)' discards qualifiers [-fpermissive]
         if (current <= AUTO_ADJUST_MAX) st.SilentStepStick2130(current);
                                                                       ^
sketch/Marlin_main.cpp:10417:53: warning: passing 'const TMC2130Stepper' as 'this' argument of 'uint16_t TMC2130Stepper::getCurrent()' discards qualifiers [-fpermissive]
         st.SilentStepStick2130((float)st.getCurrent() - CURRENT_STEP);
                                                     ^
sketch/Marlin_main.cpp:10417:69: warning: passing 'const TMC2130Stepper' as 'this' argument of 'void TMC2130Stepper::SilentStepStick2130(uint16_t)' discards qualifiers [-fpermissive]
         st.SilentStepStick2130((float)st.getCurrent() - CURRENT_STEP);
                                                                     ^

Sketch uses 55,100 bytes (21%) of program storage space. Maximum is 253,952 bytes.
Global variables use 3,271 bytes (39%) of dynamic memory, leaving 4,921 bytes for local variables. Maximum is 8,192 bytes.
```
doesn't look good.

**Comment by nesalkiN** [comment]
Great work! Thanks!

You have removed some stuff related to NO_WORKSPACE_OFFSETS?

**Comment by ghost** [comment]
Oh.. I don't know. I was trying to update this branch with the latest RcBugFix.

I have some kinks to work out.

**Comment by nesalkiN** [comment]
Ah, I see. It was added here https://github.com/MarlinFirmware/Marlin/pull/5963

**Comment by ghost** [comment]
Okay. I will check and add the missing parts.

**Comment by teemuatlut** [comment]
@Tannoo Those Travis are unrelated to this and will be fixed when I submit a new PR for updated TMC2130 settings and features.

**Comment by ghost** [comment]
Now, on to figuring out the LCD's problem.

I believe it has something to do with not saving the last color values.

**Comment by nesalkiN** [comment]
What is the lcd problem?

Testing this now. Working good printing using USB. With LEDSTRIP_EXCHANGE_RU I got strange colors while print was running. M150 G255 showed red, M150 U255 green and M150 B255 blue. Like it should. Removed LEDSTRIP_EXCHANGE_RU and M150 worked the same but the colors seems to be the right ones while printing.

Edit: Ooops, forgot my M150-lines in my start and end gcode. ;)
I have some start and end gcodes that repetier-server adds before print and after print. The end part does not run, the printer head just stops and wait for me to press a button. I have lifting and x homing there.
May need to change stuff in ledstrip.cpp I guess.

**Comment by ghost** [comment]
Yeah.. Still trying to find a good spot for that.

**Comment by ghost** [comment]
The LCD issue is the controlling of the ledstrip. The only thing that is reliable, it that it will turn them off.
It does not save the previous color. Also, the P returned value changes to 1 and stays.
The Segment numbers >0 do not work. Anything higher than 0 turns on the first segment...any subsequent S values >0 sent, turns on the second segment.

It is all jacked up. I've put debug stuff all over the place and am no closer to solving it.

**Comment by ghost** [comment]
M150 changes everything just fine.

My Z axis is grunting like crazy, but only when it's raising after homing. It does fine with a G0 Zxx.

**Comment by ghost** [comment]
In `marlin_main.cpp`:
```
      #if ENABLED(PRINTER_EVENT_LEDS)
        #if ENABLED(LEDSTRIP)
          byte r, g, b;
        #endif
        if(wait_for_heatup) {
          // Gradually change LED strip from blue to violet as bed heats up
          r = map(temp, starting_temp_b, theTarget, 0, 253);
          g = 0;
          b = 255;
          #if ENABLED(LEDSTRIP)
            #if ENABLED(DEBUG_LEDSTRIP)
              SERIAL_ECHO_START;
              SERIAL_ECHOPAIR("Red: ", r);
              SERIAL_ECHOPAIR(" Green: ", g);
              SERIAL_ECHOLNPAIR(" Blue: ", b);
            #endif
            SendColorsOnLedstrip (r, 0, 255, 0, 1);
            safe_delay(500);         <<<---------- Added this delay to get my Z to move better.
          #else
            set_led_color(r, g, b, w);
          #endif
        }
```
So, I am curious. Is there a way to stop this color change (without disabling `PRINTER_EVENT_LEDS`) until all axis have stopped? The constant calls to the library was causing my Z move (during bed heating) to start and stop very quickly until Z reached it's destination.

**Comment by ghost** [comment]
@NiklasOlofsson, the green and wait is when steppers are turned off (`M84`). Could re-arrainge the ending gcode to make the M84 the very last thing to execute.
```
inline void gcode_M18_M84() {
  if (code_seen('S')) {
    stepper_inactive_time = code_value_millis_from_seconds();
  }
  else {
    bool all_axis = !((code_seen('X')) || (code_seen('Y')) || (code_seen('Z')) || (code_seen('E')));
    if (all_axis) {
      stepper.finish_and_disable();
    }
    else {
      stepper.synchronize();
      if (code_seen('X')) disable_x();
      if (code_seen('Y')) disable_y();
      if (code_seen('Z')) disable_z();
      #if ((E0_ENABLE_PIN != X_ENABLE_PIN) && (E1_ENABLE_PIN != Y_ENABLE_PIN)) // Only enable on boards that have seperate ENABLE_PINS
        if (code_seen('E')) {
          disable_e0();
          disable_e1();
          disable_e2();
          disable_e3();
        }
      #endif
    }
  }
  #if ENABLED(PRINTER_EVENT_LEDS)
    handle_led_print_event(0);  // Turn RGB green with click to off or timeout to off
  #endif
}
```

I couldn't find a better spot to put it to know that the printer is done printing via serial or sd.

**Comment by nesalkiN** [comment]
Hmm, I send M84 in my end code but I could change the leds using M150 there too. I just don't wan't it to pause before repetier has sent the end gcodes.

**Comment by ghost** [comment]
Mine homes X and Y after a print, then sends M84.

**Comment by nesalkiN** [comment]
I could add the codes to the slicer but now repetier adds them before and after print. So, the print is done but repetier-server has a few more commands to send. If I move the commands to the slicer, I have to update all the profiles if I change anything in the start and end gcode.

**Comment by ghost** [comment]
You could disable `PRINTER_EVENT_LEDS`, and insert your `M150` commands into the slicer of choice.

**Comment by nesalkiN** [comment]
I like PRINTER_EVENT_LEDS, can't get that slow color change based on temp and stuff using M150?

**Comment by ghost** [comment]
No...that uses the arduino `map` function to do that.

**Comment by ghost** [comment]
Encapsulate this with `NO_PAUSE_OR_TIMEOUT` in `ledstrip.cpp` like so:
```
        #if DISABLED(NO_PAUSE_OR_TIMEOUT)
          wait_for_user = true;
          int wait_for_user_timeout;

          #if ENABLED(ULTRA_LCD)
            wait_for_user_timeout = 0;
            do {
             idle();
             safe_delay(100);
             if (wait_for_user_timeout == LEDSTRIP_RESET_TIME / 2)
             SERIAL_ECHOLNPGM(MSG_BUSY_PAUSED_FOR_USER_OR_TIMEOUT);
             wait_for_user_timeout ++;
             if (wait_for_user_timeout >= LEDSTRIP_RESET_TIME) break;
            } while (wait_for_user);
          #endif // ULTRA_LCD
  
          #if ENABLED(DOGLCD)
            wait_for_user_timeout = 0;
            do {
             idle();
             safe_delay(100);
             if (wait_for_user_timeout == LEDSTRIP_RESET_TIME / 2)
             SERIAL_ECHOLNPGM(MSG_BUSY_PAUSED_FOR_USER_OR_TIMEOUT);
             wait_for_user_timeout ++;
             if (wait_for_user_timeout >= LEDSTRIP_RESET_TIME) break;
            } while (wait_for_user);
          #endif // DOGLCD

          wait_for_user = false;
          SendColorsOnLedstrip (0, 0, 0, 0, 0);  // Turn RGB LEDs off
        #endif  // NO_PAUSE_OR_TIMEOUT
```

And add `#define NO_PAUSE_OR_TIMEOUT` to the Printer Event Strip Control in `configuration_adv.h` like so:
```
#if ENABLED(RGB_STRIP) || ENABLED(LEDSTRIP)
  #define PRINTER_EVENT_LEDS
  #define NO_PAUSE_OR_TIMEOUT  // Removes the pause for click or timeout on event 0 (End of print).
#endif
```
They will say on forever this way. No pauses, no clicks. And optional.

File: Marlin/Configuration.h [hunk 2]
```diff
@@ -1472,6 +1472,22 @@
...
+  //=========================== Printer Event LEDs ============================
+  //===========================================================================
+ *  When printing, the LEDs will display printer status
+ *  - LEDs will gradually change from blue to violet as the heated bed gets to target temp
+ *  - LEDs will gradually change from violet to red as the hotend gets to temperature
+ *  - LEDs will change to white to illuminate work surface
+ *  - LEDs will change to green once print has finished
+ *  - LEDs will turn off after the print has finished and the user has pushed a button 
+ * */
+#if ENABLED(BLINKM) || ENABLED(RGB_LED) || ENABLED(RGB_STRIP)
```
Comment (nesalkiN) [inline Marlin/Configuration.h]: || ENABLED(LEDSTRIP) missing here?

**Comment by nesalkiN** [comment]
I would like to get full white on when I turn on the printer. I tried the auto0.g file on SD with M150 R255 U255 B255 and it runs the file at power on. The light think that the printer is printing and finishing the print (only one line in the file) so my white does not come on before green takes over ;). How to fix?

Ah, just changed this one:

  #if ENABLED(LEDSTRIP)
    SendColorsOnLedstrip (255, 255, 255, 0, 1);  // Turn on the strip
  #endif

**Comment by nesalkiN** [comment]
Also, I have seen that the temperatures on the display does not go up as fast as they did before. The leds react all the time while it should  but the temperature freezes for a short time and then jumps upp a couple of degrees to the right value, sen freeze again for like 2-3 degrees.

**Comment by ghost** [comment]
Well, I'm still trying to find a good balance between the fastLed's library delays and Marlin. As long as Marlin doesn't loose control.

**Comment by ghost** [comment]
I will look at it more in the am.

**Comment by ghost** [comment]
Ledstrip's printer events are in the conf_adv file.

**Comment by ghost** [comment]
That is also where I moved the rgb_strip option.

Rgb_led and blinkm should also be moved there.

I plan on consolidating the printer events into one. Just haven't got that far yet.

File: Marlin/Configuration.h (hunk 2 above)
Comment (ghost) [reply Marlin/Configuration.h]: Nope. Look in `configuration_adv.h`.

**Comment by ghost** [comment]
The fasLED library just might not be fast enough for the fade to work as with a standard RGB led or strip.

**Comment by nesalkiN** [comment]
You are doing a great job. I'm trying to learn more about Arduino coding but I don't know enough to help out yet.

**Comment by ghost** [comment]
I can make an option for on or off on startup.

**Comment by ghost** [comment]
I made some changes. I added the option to be on full white on startup for the `RGB_STRIP` and the `LEDSTRIP`features.
I have also moved the `BLINKM`, `RGB_LED`, and `PRINTER_EVENT_LEDS` configuration options to the `configuration_adv` file.

**Comment by ghost** [comment]
I tried getting rid of the delay for the hotend. Well, that just caused 20 deg overshoots for me. So, I went back to the delay. Maybe @boweeble, do you have a  better idea on that?

I have also been trying to keep up with recent merges to RCBugFix.

**Comment by nesalkiN** [comment]
Changed G28 leds from off to white. I do G28 before print or just to park my nozzle in the middle. Need the leds on. I hope that doesn't affect anything. ;)

**Comment by ghost** [comment]
> Changed G28 leds from off to white. I do G28 before print or just to park my nozzle in the middle. Need the leds on. I hope that doesn't affect anything. ;)

Nope, you can do what you want with the code you are running.
Although, I did make a new option to do just that:
```
//#define LIGHT_ON_POWERUP      // Light the entire strip WHITE on startup.
```
It does this:
```
  #if ENABLED(LEDSTRIP) && ENABLED(LIGHT_ON_POWERUP)
    SendColorsOnLedstrip (255, 255, 255, 0, 1);  // Turn on the strip
  #elif ENABLED(LEDSTRIP)
    SendColorsOnLedstrip (0, 0, 0, 0, 0);  // Turn off the strip
  #endif
```
This gives others the same option by just defining `LIGHT_ON_POWERUP`.

**Comment by nesalkiN** [comment]
When I power up the printer it get white light which is perfect. Later, I send G28 (manually with G28 or home all in repetier). The leds turns yellow/green and when G28 is done the leds turned off.  Not always starting the print after doing that. ;)

**Comment by ghost** [comment]
Oh...Yeah, that's a different event. Yeah, I set yellow for for homing.

That can be changed in `main_marlin.cpp` LN:3643,
```
  #if ENABLED(PRINTER_EVENT_LEDS)
    handle_led_print_event(9);  // Turn RGB LEDs off
  #endif
```
Change the `9` to `1`.

**Comment by ghost** [comment]
Here are some other "whites" you can use:
```
        /* OTHER COLORS BY NAME AND VALUE FOR REFERENCE
         *        COLOR NAME           R   G(U)  B
         * Candle                     255, 147, 41
         * 40W Tungsten               255, 197, 143
         * 100W Tungsten              255, 214, 170
         * Halogen                    255, 241, 224
         * Carbon Arc                 255, 250, 244
         * High Noon Sun              255, 255, 251
         * Direct Sunlight            255, 255, 255
         * Overcast Sky               201, 226, 255
         * Clear Blue Sky             64, 156, 255
         * Warm Fluorescent           255, 244, 229
         * Standard Fluorescent       244, 255, 250
         * Cool White Fluorescent     212, 235, 255
         * Full Spectrum Fluorescent  255, 244, 242
         * Grow Light Fluorescent     255, 239, 247
         * 
         *      OTHER COLORS
         * Black Light Fluorescent    167, 0, 255
         */
```
I grabbed these just in case you didn't want the 'Direct Sunlight' on your printer. :)

**Comment by ghost** [comment]
Yeah...Something went screwy again. I'll get this later.

Well, it is later....it was nagging me. lol

**Comment by Roxy-3D** [comment]
Wow!   That is a lot of interaction with GitHub!   If Travis starts cutting you any slack, people are going to start talking...

**Comment by ghost** [comment]
I wish I could squash without messing up my repo.

I will likely close the PR and submit a new one when new features are welcome.

**Comment by ghost** [comment]
For now, I'm just trying to keep it up to date.

**Comment by teemuatlut** [comment]
Can you make another branch that's squashed and with what you make the PR, while you also keep this branch for yourself for development? I've actually been waiting for this to get merged.

**Comment by ghost** [comment]
I can make another PR, but with all the changes happening with RCBugFix, it would make a ton of commits there also. I also don't think it will be merged until the next release is pushed out.
I also don't have the LCD part quite figured out yet. It worked on the version I got it from, but not really in this release.

**Comment by thinkyhead** [comment]
I picked up on the original work by @boweeble and initiated a new PR #6240, which you may add to by checking out the branch, modifying it, and posting a PR to my fork. The branch in question is https://github.com/thinkyhead/Marlin/tree/rebuild_rgb_led. Make a copy using the console in Github Desktop…
```
git remote add thinkyhead git@github.com:thinkyhead/Marlin.git
git fetch thinkyhead
git checkout thinkyhead/rebuild_rgb_led -b my_rgb_led_work
git push --set-upstream origin my_rgb_led_work
```

Or if you prefer, I can finish up the basic RGB implementation, merge it, and then you can start working from `RCBugFix`.

**Pull request closed without merging**